lib/Modules/CinderHooks/CinderHooks.psm1

```diff
@@ -283,11 +347,88 @@ function Get-CharmConfigContext {
     return $ctxt
 }
 
+function Get-CertificatesContext {
+    Write-JujuWarning "Generating context for tls-certificates"
+
+    $requiredCtx = @{
+        "ca" = $null
+    }
+
+    $optionalCtx = @{
+        "client.cert" = $null
+        "client.key" = $null
+    }
+
+    $ctxt = Get-JujuRelationContext -Relation "certificates" `
+                                    -OptionalContext $optionalCtx `
+                                    -RequiredContext $requiredCtx
+
+    if (!$ctxt.Count) {
+        return @{}
+    }
+
+    Set-Content $CINDER_CA_CERT $ctxt["ca"]
+    $ctxt["ssl_ca_cert"] = $CINDER_CA_CERT
+
+    return $ctxt
+}
+
+function Get-KeystoneCredentialsContext {
+    Write-JujuWarning "Generating context for keystone credentials"
+    $requiredCtx = @{
+        "credentials_project" = $null
+        "credentials_project_id" = $null
+        "credentials_host" = $null
+        "credentials_port" = $null
+        "credentials_protocol" = $null
+        "credentials_username" = $null
+        "credentials_password" = $null
+        "credentials_project_domain_name" = $null
+        "credentials_user_domain_name" = $null
+        "region" = $null
+        "api_version" = $null
+    }
+
+    $optionalCtx = @{
+        "ca_cert" = $null
+    }
+    $ctxt = Get-JujuRelationContext -Relation "keystone" `
+                                    -OptionalContext $optionalCtx `
+                                    -RequiredContext $requiredCtx
+
+    if (!$ctxt.Count) {
+        return @{}
+    }
+
+    if (!$ctxt["api_version"] -or $ctxt["api_version"] -eq 2) {
+        $ctxt["api_version"] = "2.0"
+    }
+
+    $authurl = "{0}://{1}:{2}/v{3}/" -f @(
+                            $ctxt['credentials_protocol'],
+                            $ctxt['credentials_host'],
+                            $ctxt['credentials_port'],
+                            $ctxt['api_version']
+                        )
+
+    $ctxt["admin_auth_url"] = $authurl
+
+    $ksCa = Join-Path $CINDER_INSTALL_DIR "etc\keystone_ca_cert.pem"
+
+    if ($ctxt["ca_cert"]) {
+        Write-FileFromBase64 -File $ksCa -Content $ctxt["ca_cert"]
+        $ctxt["ks_ssl_ca_cert"] = $ksCa
+    }
+
+    return $ctxt
+}
+
 function Get-CloudComputeContext {
```

Comment:
We don't have the `cloud-compute` provides relation anymore.

Do we still need this function ?

Reply:
nope!